crud/borders.lua

```diff
@@ -0,0 +1,180 @@
+
+
+local function get_border_on_storage(border_name, space_name, index_id, field_names)
+    dev_checks('string', 'string', 'number', '?table')
+
+    assert(border_name == 'min' or border_name == 'max')
+
+    local space = box.space[space_name]
+    if space == nil then
+        return nil, BorderError:new("Space %q doesn't exist", space_name)
+    end
+
+    local index = space.index[index_id]
+    if index == nil then
+        return nil, BorderError:new("Index %q of space doesn't exist", index_id, space_name)
+    end
+
+    local function get_border(index)
+        return index[border_name](index)
+    end
+
+    return schema.wrap_func_result(space, get_border, {index}, {
+        add_space_schema_hash = true,
+        field_names = field_names,
+    })
+end
+
+function borders.init()
+   _G._crud.get_border_on_storage = get_border_on_storage
+end
+
+local function is_closer(border_name, keydef, tuple, res_tuple)
+    assert(border_name == 'min' or border_name == 'max')
+
+    if res_tuple == nil then
+        return true
+    end
+
+    local cmp = keydef:compare(tuple, res_tuple)
+
+    return border_name == 'min' and cmp < 0 or border_name == 'max' and cmp > 0
+end
+
+local function get_border(border_name, space_name, index_name, opts)
+    checks('string', 'string', '?string|number', {
+        timeout = '?number',
+        fields = '?table',
+    })
+
+    opts = opts or {}
+
+    local space = utils.get_space(space_name, vshard.router.routeall())
+    if space == nil then
+        return nil, BorderError:new("Space %q doesn't exist", space_name), true
+    end
+
+    local index
+    if index_name == nil then
+        index = space.index[0]
+    else
+        index = space.index[index_name]
+    end
+
+    if index == nil then
+        return nil, BorderError:new("Index %q of space %q doesn't exist", index_name, space_name)
+    end
+
+    local primary_index = space.index[0]
+
+    local cmp_key_parts = utils.merge_primary_key_parts(index.parts, primary_index.parts)
+    local field_names = utils.enrich_field_names_with_cmp_key(opts.fields, cmp_key_parts, space:format())
+
+    local replicasets = vshard.router.routeall()
+    local call_opts = {
+        mode = 'read',
+        replicasets = replicasets,
+        timeout = opts.timeout,
+    }
+    local results, err = call.map(
+        STAT_FUNC_NAME,
+        {border_name, space_name, index.id, field_names},
+        call_opts
+    )
+
+    if err ~= nil then
+        return nil, BorderError:new("Failed to get %s: %s", border_name, err)
+    end
+
+    local keydef = Keydef.new(space, field_names, index.id)
+
+    local tuples = {}
+    for _, result in pairs(results) do
+        if result[1].err ~= nil then
+            return nil, BorderError:new("Failed to get %s: %s", border_name, result.err)
+        end
+
+        local tuple = result[1].res
+        if tuple ~= nil then
+            table.insert(tuples, tuple)
+        end
+    end
```

Comment:
I suppose you can calculate min/max without additional loop and temporary `tuples` table.